### ISSUE TITLE
Video comments and likes design brief

### ISSUE THREAD

**Issue by bedeho** [Joystream/atlas#1812]
I was going to do do a full brief, but doing something brief like this felt more productive, since the comment & reaction form factor is so familiar.

**Reactions:**

Let us briefly clarify what exactly is meant by this. This means that a given entity, such a  a video or a comment, allows any member to assign (and later unassign if desired) one among a finite set of identifiable markers to the entity. The set of possible markers (like, unlike, shocked, etc.) is public.

**Tech:**

* An early version of this has been implemented in the runtime (Ignazio), however there is a substantial likelihood we will redo this as a metaprotocol, instead of holing comment state in runtime storage. The repeating issues are complexity around fees for managing state bloat, and the problems associated with lack of on-chain iteration, which really is needed for this sort of state (e.g. delete all comments). This means that while implementing this, the tech team should just ignore all the state bloat bonds required in extrinsics, and also ignore the need to clean up such bonds later. This may sound complex, but it is easy to figure out when looking at the runtime extrinsics API. Lastly, whenever we make the change from a runtime to a metaprotocol version, Atlas should largely be unaffected, as the query node API will be largely unaffected, and extrinsics will only be superficially impacted.
* There is no query node support at the moment.
* The implementation in the runtime is semantically accurate, that is:
  * each video has a threaded comment section by default.
  * each video has a set of reactions: same semantics as e.g. FB.
  * a membership is needed to create a comment.
  * each comment has a set of reactions: same semantics as e.g. FB.
  * the comment creator can edit their comment later.
  * a comment can be deleted either by the commenter or the channel, however descendant comments still remain.
  * the owner of the channel can turn the comment section on or off at any time, and when off, all efforts to update comments, or reactions to them, are ignored.
  * the owner of the channel can do the same with the video level reactions.
  * **not supported but probably should be for this epic:**
    * a channel owner should be able to ban a specific other member from participating on any video.
    * a channel owner can signal that a given comment is pinned on a video.
* As the design section mentions, notifications seem very important here, however, without a server side storage it can get hard to show accurate set of notifications as the number of comments start getting into the hundreds or thousands. Unclear to me what the best path is here.

Just to be clear, all _actions_ mentioned above are _extrinsics_.

There is a non-negligible chance that my description of _what is_ is slightly off, so I will ask @ignazio-bovo to review and give feedback here.

**Design:**

* There is an open question here about whether we should keep the very heavy & blocking interaction style we use when doing all extrinsics so far, where we block the view of the user until both the extrinsic as been finalized and the query node has caught up with the block where the extrinsic is included. All the the prior actions we have worked on have been very high payoff for the actor, hence the overhead was more acceptable, but when we are now venturing into more open ended social interactions, like reacting and commenting, this may become a burden too heavy to bare. An alternative approach would be to allow the user to fire off extrinsics, and then to not block the view, and instead display some sort of user-side loading state in some local part of the view. This would allow the user to start doing other things, or even go to another screen, and the final outcome could be signaled with a success alert or something. The flip side of this is that it would be new and perhaps risky in some way to execute on. More discussion is needed.
* Its not clear how comments should be ranked, other than the creator pinned comment on top. There are at least two distinct metrics of relevance: reaction count and creation date. But how should they be combined into one coherent ranking?
* Remember that a member who has reacted should be able to un-react.
* Notifications will be key here:
  * Some one with a popular channel risks getting in inundated with massive volumes of comment notifications, while for a new up-and coming creator, getting notified about each and every comment may be essential for community development and positive reinforcement. Certainly some level of control about what notifications to show must be a setting that can be set per video and perhaps across the channel.
  * As someone who makes a comment, I am certainly also interested and learning that other people are making replies to my comment or reacting to it.
* While we allow comment editing, and should always show the latest version, it should be possible to know that something was edited, and go back to prior edits.

**Strategic**

Perhaps there is scope to restrict the scope a bit here, because everything mentioned seems quite big for a first iteration, e.g. notifications? Discuss.

┆Issue is synchronized with this [Asana task](https://app.asana.com/0/1201958687417145/1201959043907249) by [Unito](https://www.unito.io)

**Comment by dobertRowneySr** [comment]
I will go over the statements one by one:
- There is no query node support at the moment :+1: 
- each video has a threaded comment section by default :-1:  
   1. *A video can have a optional description (i.e. video post id, see details section below)*
   ``` rust
   struct VideoRecord  {
       video_post_id: Option<PostId>,
       ...
   }
   ```
   2. a video can be uploaded and no threaded comment section must be necessarily created
  3. a threaded section is completely removed by deleting the `video_post_id` for that video
- each video has a set of reactions: same semantics as e.g. FB :+1: 
 *`ReactionId` is a `NumericIdentifier` (`u64` basically) and I have not enriched the type to enforce the semantics that the possible value should belong to a finite set (since we want a finite number of reactions). The reaction set has not been be defined yet*
- a membership is needed to create a comment :+1: 
- each comment has a set of reactions: same semantics as e.g. FB :+1: 
*Same answer as video reaction*
- the comment creator can edit their comment later :+1: 
- a comment can be deleted either by the commenter or the channel, however descendant comments still remain :+1: *This is true for `PostType::Comment(_)`, while deleting a 
`PostType::VideoPost` will delete the whole threaded comment section. There is also a moderator set for each channel authorized to moderate the comment section. The latter is rip-off the olympia forum module*
-  the owner of the channel can turn the comment section on or off at any time, and when off, all efforts to update comments, or reactions to them, are ignored :+1: 
- the owner of the channel can do the same with the video level reactions :-1: 
*This is not implemented yet, but can be easily done*
- all action above are extrinsics :-1:  *comment section can be disabled/enabled only at video creation, I chose this design because I believed that it was more similar to YT*
## Details
- A Threaded Video comment section is structured as a tree:
    ``` rust
    enum PostType<PostId> {
        Comment(PostId),
         VideoPost,
    }
    ```
- `VideoPost` is the root node, whose semantics is equivalent to a YT video descriptions.
`Comment(PostId)` is a reply to a post identified by `PostId`. A comment can be on the main thread if `PostId` refers to a `VideoPost` or a reply to a previous comment
With each comment referring to the parent comment 
- Reaction are not saved on chain
- Moderator set is created as a set of valid joystream members. when a moderator deletes a comment he must provide a rationale

**Comment by kdembler** [comment]
Had a backlog of things to look at and just getting to this, have some questions:

1 .@bedeho, you wrote `there is a substantial likelihood we will redo this as a metaprotocol`. If this is the case, when would that work commence and who would do that? I fail to understand what the implications are, you're saying API would be affected only superficially?
2. @ignazio-bovo seems the `VideoPost` is the root of all the comments for a given video, right? You wrote `VideoPost is the root node, whose semantics is equivalent to a YT video descriptions.` What semantics exactly are that? Just a string with the description? Videos already have description in their metadata, so the content of `VideoPost` would likely be ignored in the UI. 
3. @ignazio-bovo Would `VideoPost` be automatically created on video creation if the comments are enabled?
4. @ignazio-bovo on reactions to comments and videos, is there any logic that validates reactions? For example, can runtime restrict number of reactions a given user can send to a video? Assuming we want to do likes/dislikes on videos, can we stop user from liking a single video X number of times?
5. @ignazio-bovo in `Details` section you wrote `Reaction are not saved on chain`, what does that refer to?

**Comment by bedeho** [comment]
1.
> If this is the case, when would that work commence and who would do that?

I think the earliest this could commence would be when either @ondratra is done with his Olympia tasks, or @zeeshanakram3 is done with his current merkle tree based cashout work for creators. So that is likely at least a week or two from now.

> I fail to understand what the implications are, you're saying API would be affected only superficially?

Yes, very much so, as long as you ignore the state bloat bond creation and reclaiming.

**Comment by dobertRowneySr** [comment]
2. Yes `Video.video_post_id` is the root of the comment section, it was meant to be *exactly* like the metadata you mentioned about, but it was one of the first projects that I did and I didn't know much about metadata. The code can be cleaned up if necessary
3. `Video.video_post_id` is optional upon creation (i.e. it can be `None`), it can only be added/modified by the channel owner, which is the only one authorized also to enable the comment section. The only problem is having comments enabled and `video_post_id == None` and that would not  allow you to post comment since you would not have a `parent_id` to specify. I realize now  that is a (small?) bug @bedeho. So that can be a strong signal to simplify (as mentioned in (2))
5. `react_to_post(member_id, reaction_id, post_id, video_id)` just checks validity of `member_id` and deposits event for `reaction` no extra operation are performed and there is no element in `decl_storage!` to keep track of reaction in the content pallet state.
4. You can react unlimited number of times. If you want to limit one member to have <= 1 reaction you would need an extra `StorageMap` `MemberId => ReactionId` in the content pallet state to keep track of the reactions by each user so far.

**Comment by kdembler** [comment]
All right, now it's all clear, thanks!

To followup:

2 + 3. I think it would make sense to clean this up slightly:
a) all the metadata stuff is already handled so we could entirely skip contents of `VideoPost`
b) It'd be great to create `VideoPost` automatically if comments are enabled, it doesn't make much sense to have comments enabled but the `VideoPost` missing and I think we would like to avoid publisher signing a separate extrinsic just to enable the comments

4 + 5. I would argue that this is important to have, I think we shouldn't let users leave arbitrary number of likes under a single video. WDYT @bedeho?

**Comment by bedeho** [comment]
> 4 + 5. I would argue that this is important to have, I think we shouldn't let users leave arbitrary number of likes under a single video. WDYT @bedeho?

I could see the case for both, unlimited number of likes (as Medium does), also can work well, but in our case the primary thing for Atlas should be the familiar one-member-one-interaction. The cost of adding such a `MemberId => ReactionId` map in the runtime is one of the many reasons we will redo this as a metaprotocol. I think you should do the designs as if its one-member-one-interaction.

**Comment by kdembler** [comment]
Just noticed that runtime atm doesn't let you un-react, once you post a reaction, you cannot remove it, both on video and comment. We should add this functionality. Just dropping here so we don't forget

**Comment by bedeho** [comment]
> Just noticed that runtime atm doesn't let you un-react, once you post a reaction, you cannot remove it, both on video and comment. We should add this functionality. Just dropping here so we don't forget

We will def. forget if this comment is all that becomes of it 🤣 , added a note here => https://github.com/Joystream/joystream/issues/2932

**Comment by dmtrjsg** [comment]
Implemented and released to production #1876 #1986
